File: rust/features/steps/initialization_steps.rs
use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;
use std::process::Command;
use std::time::SystemTime;
use std::thread::JoinHandle;

use cucumber::{given, then, when, World};
use tempfile::TempDir;

use taskulus::cli::run_from_args_with_output;
use taskulus::index::IssueIndex;
use taskulus::models::ProjectConfiguration;
use taskulus::daemon_client;

#[derive(Debug, Default, World)]
pub struct TaskulusWorld {
    pub temp_dir: Option<TempDir>,
    pub working_directory: Option<PathBuf>,
    pub exit_code: Option<i32>,
    pub stdout: Option<String>,
    pub stderr: Option<String>,
    pub configuration: Option<ProjectConfiguration>,
    pub generated_id: Option<String>,
    pub generated_ids: Option<HashSet<String>>,
    pub id_prefix: Option<String>,
    pub existing_ids: Option<HashSet<String>>,
    pub project_dirs: Option<Vec<PathBuf>>,
    pub project_error: Option<String>,
    pub cache_path: Option<PathBuf>,
    pub cache_mtime: Option<SystemTime>,
    pub daemon_spawned: bool,
    pub daemon_connected: bool,
    pub stale_socket_removed: bool,
    pub daemon_rebuilt_index: bool,
    pub daemon_simulation: bool,
    pub protocol_errors: Vec<String>,
    pub protocol_error: Option<String>,
    pub daemon_response_code: Option<String>,
    pub daemon_error_message: Option<String>,
    pub daemon_index_issues: Option<Vec<String>>,
    pub daemon_spawn_called: bool,
    pub daemon_entry_running: bool,
    pub daemon_list_error: bool,
    pub local_listing_error: bool,
    pub shared_only_results: Option<Vec<String>>,
    pub expected_project_dir: Option<PathBuf>,
    pub expected_project_path: Option<PathBuf>,
    pub force_empty_projects: bool,
    pub hash_sequence: Option<Vec<String>>,
    pub random_bytes_override: Option<String>,
    pub migration_errors: Vec<String>,
    pub workflow_error: Option<String>,
    pub index: Option<IssueIndex>,
    pub daemon_thread: Option<JoinHandle<()>>,
    pub daemon_fake_server: bool,
    pub daemon_mode_disabled: bool,
    pub current_user: Option<String>,
    pub original_taskulus_user: Option<Option<String>>,
    pub original_user_env: Option<Option<String>>,
    pub formatted_output: Option<String>,
}

impl Drop for TaskulusWorld {
    fn drop(&mut self) {
        if let Some(handle) = self.daemon_thread.take() {
            if !self.daemon_fake_server {
                if let Some(root) = self.working_directory.as_ref() {
                    let _ = daemon_client::request_shutdown(root);
                }
            }
            let _ = handle.join();
        }
        if let Some(original) = self.original_taskulus_user.take() {
            match original {
                Some(value) => std::env::set_var("TASKULUS_USER", value),
                None => std::env::remove_var("TASKULUS_USER"),
            }
        }
        if let Some(original) = self.original_user_env.take() {
            match original {
                Some(value) => std::env::set_var("USER", value),
                None => std::env::remove_var("USER"),
            }
        }
    }
}

fn run_cli(world: &mut TaskulusWorld, command: &str) {
    let args = shell_words::split(command).expect("parse command");
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");

    match run_from_args_with_output(args, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[given("an empty git repository")]
fn given_empty_git_repository(world: &mut TaskulusWorld) {
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a directory that is not a git repository")]
fn given_directory_not_git_repository(world: &mut TaskulusWorld) {
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("not-a-repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a git repository with an existing Taskulus project")]
fn given_existing_taskulus_project(world: &mut TaskulusWorld) {
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("existing");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    fs::create_dir_all(repo_path.join("project").join("issues")).expect("create project");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a git repository metadata directory")]
fn given_git_metadata_directory(world: &mut TaskulusWorld) {
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("metadata");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.join(".git"));
    world.temp_dir = Some(temp_dir);
}

#[when("I run \"tsk init\"")]
fn when_run_tsk_init(world: &mut TaskulusWorld) {
    run_cli(world, "tsk init");
}

#[when("I run \"tsk init --local\"")]
fn when_run_tsk_init_local(world: &mut TaskulusWorld) {
    run_cli(world, "tsk init --local");
}

#[then("a \".taskulus.yaml\" file should not exist")]
fn then_marker_missing(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    assert!(!cwd.join(".taskulus.yaml").exists());
}

#[then("a \"project\" directory should exist")]
fn then_project_directory_exists(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    assert!(cwd.join("project").is_dir());
}

#[then("a \"project/config.yaml\" file should not exist")]
fn then_default_config_missing(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    assert!(!cwd.join("project").join("config.yaml").exists());
}

#[then("a \"project/issues\" directory should exist and be empty")]
fn then_issues_directory_empty(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    let issues_dir = cwd.join("project").join("issues");
    assert!(issues_dir.is_dir());
    assert!(issues_dir
        .read_dir()
        .expect("read issues dir")
        .next()
        .is_none());
}

#[then("a \"project/wiki\" directory should not exist")]
fn then_wiki_directory_missing(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    assert!(!cwd.join("project").join("wiki").exists());
}

#[then("a \"project-local/issues\" directory should exist")]
fn then_local_issues_directory_exists(world: &mut TaskulusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    assert!(cwd.join("project-local").join("issues").is_dir());
}

#[then("the command should fail with exit code 1")]
fn then_command_failed(world: &mut TaskulusWorld) {
    assert_eq!(world.exit_code, Some(1));
}

#[then("the command should fail")]
fn then_command_failed_generic(world: &mut TaskulusWorld) {
    assert_ne!(world.exit_code, Some(0));
}

File: python/features/steps/initialization_steps.py
"""Behave steps for project initialization."""

from __future__ import annotations

from pathlib import Path
from behave import given, then

from features.steps.shared import ensure_git_repository


@given("an empty git repository")
def given_empty_git_repository(context: object) -> None:
    repository_path = Path(context.temp_dir) / "repo"
    repository_path.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(repository_path)
    context.working_directory = repository_path


@given("a directory that is not a git repository")
def given_directory_not_git_repository(context: object) -> None:
    repository_path = Path(context.temp_dir) / "not-a-repo"
    repository_path.mkdir(parents=True, exist_ok=True)
    context.working_directory = repository_path


@given("a git repository with an existing Kanbus project")
def given_existing_kanbus_project(context: object) -> None:
    repository_path = Path(context.temp_dir) / "existing"
    repository_path.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(repository_path)
    (repository_path / "project" / "issues").mkdir(parents=True)
    context.working_directory = repository_path


@given("a git repository metadata directory")
def given_git_metadata_directory(context: object) -> None:
    repository_path = Path(context.temp_dir) / "metadata"
    repository_path.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(repository_path)
    context.working_directory = repository_path / ".git"


@then('a ".kanbus.yml" file should be created')
def then_marker_created(context: object) -> None:
    assert (context.working_directory / ".kanbus.yml").is_file()


@then('a "CONTRIBUTING_AGENT.template.md" file should be created')
def then_project_management_template_created(context: object) -> None:
    assert (context.working_directory / "CONTRIBUTING_AGENT.template.md").is_file()


@then('CONTRIBUTING_AGENT.template.md should contain "{text}"')
def then_project_management_template_contains_text(context: object, text: str) -> None:
    normalized = text.replace('\\"', '"')
    content = (context.working_directory / "CONTRIBUTING_AGENT.template.md").read_text(
        encoding="utf-8"
    )
    assert normalized in content


@then('a "project" directory should exist')
def then_project_directory_exists(context: object) -> None:
    assert (context.working_directory / "project").is_dir()


@then('a "project/config.yaml" file should not exist')
def then_default_config_missing(context: object) -> None:
    assert not (context.working_directory / "project" / "config.yaml").exists()


@then('a "project/issues" directory should exist and contain only guard files')
def then_issues_directory_only_guards(context: object) -> None:
    issues_dir = context.working_directory / "project" / "issues"
    assert issues_dir.is_dir()
    names = {p.name for p in issues_dir.iterdir()}
    assert names == {
        "AGENTS.md",
        "DO_NOT_EDIT",
    }, f"expected only guard files, got {names}"


@then('the file "{path}" should exist')
def then_file_exists(context: object, path: str) -> None:
    full_path = context.working_directory / path
    assert full_path.exists(), f"expected file {path} to exist"
    assert full_path.is_file(), f"expected {path} to be a file"


@then('the file "{path}" should contain "{text}"')
def then_file_contains(context: object, path: str, text: str) -> None:
    full_path = context.working_directory / path
    content = full_path.read_text(encoding="utf-8")
    assert text in content, f"expected {path} to contain {text!r}"


@then('a "project-local/issues" directory should exist')
def then_local_issues_directory_exists(context: object) -> None:
    assert (context.working_directory / "project-local" / "issues").is_dir()


@then("the command should fail with exit code 1")
def then_command_failed(context: object) -> None:
    if context.result.exit_code != 1:
        print(f"Expected exit code 1, got {context.result.exit_code}")
        print(f"STDOUT: {context.result.stdout}")
        print(f"STDERR: {context.result.stderr}")
    assert context.result.exit_code == 1


@then("the command should fail")
def then_command_failed_generic(context: object) -> None:
    assert context.result.exit_code != 0


@then("project/issues/ and project/events/ should contain AGENTS.md with the warning")
def then_issues_events_agents_created(context: object) -> None:
    for subdir in ("issues", "events"):
        path = context.working_directory / "project" / subdir / "AGENTS.md"
        assert path.is_file(), f"expected {path}"
        content = path.read_text(encoding="utf-8")
        assert "DO NOT EDIT HERE" in content
        assert "The Way" in content or "Kanbus" in content


@then("project/issues/ and project/events/ should contain DO_NOT_EDIT with the warning")
def then_issues_events_do_not_edit_created(context: object) -> None:
    for subdir in ("issues", "events"):
        path = context.working_directory / "project" / subdir / "DO_NOT_EDIT"
        assert path.is_file(), f"expected {path}"
        content = path.read_text(encoding="utf-8")
        assert "DO NOT EDIT" in content
        assert "The Way" in content
